network/src/protocols/rpc/mod.rs

```diff
@@ -169,85 +182,197 @@ pub struct Rpc<TSubstream> {
-{
-    match req {
-        RpcRequest::SendRpc(req) => {
-            let protocol = req.protocol;
-            let req_data = req.data;
-            let mut res_tx = req.res_tx;
-            let timeout = req.timeout;
-            let peer_id_str = peer_handle.peer_id().short_str();
-            // Future to run the actual outbound rpc protocol and get the results.
+    /// Handle a new inbound message. If the message is a response to a pending outbound RPC, send
+    /// response to upstream client and return immediately. Otherwise, it is an inbound rpc
+    /// request. In that case, we spawn the request in a new task and return.
+    fn handle_inbound_message(
+        &mut self,
+        notif: PeerNotification,
+        inbound_rpc_tasks: &mut FuturesUnordered<BoxFuture<'static, ()>>,
+    ) {
+        let notification_tx = self.rpc_handler_tx.clone();
+        let peer_handle = self.peer_handle.clone();
+        let peer_id_str = peer_handle.peer_id().short_str();
+        match notif {
+            PeerNotification::NewMessage(message) => {
+                match message {
+                    // This is a response to a pending outbound RPC.
+                    NetworkMessage::RpcResponse(response) => {
+                        if let Some(response_tx) =
+                            self.pending_outbound_rpcs.remove(&response.request_id)
+                        {
+                            if let Err(e) = response_tx.send(response) {
+                                // TODO: add ability to log protocol id as well?
+                                warn!(
+                                    "Failed to handle inbount RPC response from peer: {}. Error: {:?}",
+                                    peer_id_str,
+                                    e
+                                );
+                            }
+                        } else {
+                            info!(
+                                "Received response for expired request from {:?}. Discarding.",
+                                peer_id_str
+                            )
+                        }
+                    }
+                    // This is a new inbound RPC request.
+                    NetworkMessage::RpcRequest(request) => {
+                        if inbound_rpc_tasks.len() as u32 == self.max_concurrent_inbound_rpcs {
+                            // Log any errors.
+                            counters::LIBRA_NETWORK_RPC_MESSAGES
+                                .with_label_values(&["response", "declined"])
+                                .inc();
+                            warn!(
+                                "Pending inbound RPCs are at limit ({}). Not processing new inbound rpc requests",
+                                self.max_concurrent_inbound_rpcs
+                            );
+                            return;
+                        }
+                        let timeout = self.inbound_rpc_timeout;
```

Comment:
honestly, it seems kinda weird that an inbound request would timeout... like this feels like we either have a leak or services are taking too long to compute. I could understand the remote side saying this is our TTL and at which point, we can at least not send back the response... I believe this is a legacy element, but I'd like to dig into this more in a parallel track.

Reply:
I think we'll be able to remove this once we get rid of yamux - at present, it is possible that the remote does not accept our response (by not issuing a yamux window update) and we effectively "leak" the task driving the inbound RPC. So your assessment that this is vestigial of the stream-based implementation.